indra/newview/llinventorylistener.cpp

```diff
@@ -0,0 +1,276 @@
+        "[\"name\"] is a substring of object's name,\n"
+        "[\"desc\"] is a substring of object's description,\n"
+        "asset [\"type\"] corresponds to the object's asset type\n"
+        "[\"item_limit\"] sets item count limit in reply, maximum and default is 100\n"
+        "[\"filter_links\"]: EXCLUDE_LINKS - don't show links, ONLY_LINKS - only show links, INCLUDE_LINKS - show links too (default)",
+        &LLInventoryListener::collectDescendentsIf,
+        llsd::map("folder_id", LLSD(), "reply", LLSD()));
+ }
+
+
+void add_item_info(LLEventAPI::Response& response, LLViewerInventoryItem* item)
+{
+    response["items"].insert(item->getUUID().asString(),
+                             llsd::map("name", item->getName(), "parent_id", item->getParentUUID(), "desc", item->getDescription(),
+                                       "inv_type", LLInventoryType::lookup(item->getInventoryType()), "asset_type", LLAssetType::lookup(item->getType()),
+                                       "creation_date", (S32) item->getCreationDate(), "asset_id", item->getAssetUUID(), 
+                                       "is_link", item->getIsLinkType(), "linked_id", item->getLinkedUUID()));
+}
+
+void add_cat_info(LLEventAPI::Response &response, LLViewerInventoryCategory *cat)
+{
+    response["categories"].insert(cat->getUUID().asString(),
+                                  llsd::map("name", cat->getName(), "parent_id", cat->getParentUUID(), "type", LLFolderType::lookup(cat->getPreferredType())));
+}
+
+void LLInventoryListener::getItemsInfo(LLSD const &data)
+{
+    Response response(LLSD(), data);
+
+    uuid_vec_t ids = LLSDParam<uuid_vec_t>(data["items_id"]);
+    for (auto &it : ids)
+    {
+        LLViewerInventoryItem* item = gInventory.getItem(it);
+        if (item)
+        {
+            add_item_info(response, item);
+        }
+        LLViewerInventoryCategory* cat = gInventory.getCategory(it);
+        if (cat)
+        {
+            add_cat_info(response, cat);
+        }
+    }
+}
+
+void LLInventoryListener::getFolderTypeNames(LLSD const &data)
+{
+    Response response(llsd::map("type_names", LLFolderType::getTypeNames()), data);
+}
+
+void LLInventoryListener::getAssetTypeNames(LLSD const &data)
+{
+    Response response(llsd::map("type_names", LLAssetType::getTypeNames()), data);
+}
+
+void LLInventoryListener::getBasicFolderID(LLSD const &data)
+{
+    Response response(llsd::map("id", gInventory.findCategoryUUIDForType(LLFolderType::lookup(data["ft_name"].asString()))), data);
+}
+
+
+void LLInventoryListener::getDirectDescendents(LLSD const &data)
+{
+    Response response(LLSD(), data);
+    LLInventoryModel::cat_array_t* cats; 
+    LLInventoryModel::item_array_t* items;
+    gInventory.getDirectDescendentsOf(data["folder_id"], cats, items);
+
+    LLInventoryModel::item_array_t items_copy = *items;
+    for (LLInventoryModel::item_array_t::iterator iter = items_copy.begin(); iter != items_copy.end(); iter++)
```

Comment:
I was too cautious when reading a comment for `gInventory.getDirectDescendentsOf`, but since we are not modifying anything copy is not needed.

Reply:
I like `add_objects_info()`. Interestingly, since `add_objects_info()` accepts `cat_array` and `item_array` by value, we end up copying the arrays anway.

Reply:
Yep, I should probably change it.